src/azul/service/responseobjects/storage_service.py

```diff
@@ -39,12 +39,162 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
+
+    This class is to facilitate multipart upload to S3 storage. The class itself
+    is also a context manager.
+
+    Here is the sample usage.
+
+    .. code-block:: python
+
+       with MultipartUploadHandler('samples.txt', 'text/plain'):
+           handler.push(b'abc')
+           handler.push(b'defg')
+           # ...
+
+    where the context pattern will automatically shutdown upon exiting the context.
+    """
+
+    def __init__(self, object_key, content_type: str):
+        self.bucket_name = config.s3_bucket
+        self.object_key = object_key
+        self.upload_id = None
+        self.mp_upload = None
+        self.next_part_number = 1
+        self.content_type = content_type
+        self.parts = []
+        self.futures = []
+        self.thread_pool = None
+
+    def __enter__(self):
+        api_response = boto3.client('s3').create_multipart_upload(Bucket=self.bucket_name,
+                                                                  Key=self.object_key,
+                                                                  ContentType=self.content_type)
+        self.upload_id = api_response['UploadId']
+        self.mp_upload = boto3.resource('s3').MultipartUpload(self.bucket_name, self.object_key, self.upload_id)
+        self.thread_pool = ThreadPoolExecutor(max_workers=MULTIPART_UPLOAD_MAX_WORKERS)
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if type:
+            self.abort()
+            raise UnexpectedMultipartUploadAbort(f'{self.bucket_name}/{self.object_key}')
+        self.complete()
+
+    @property
+    def is_active(self):
+        return self.mp_upload is not None
+
+    def complete(self):
+        """
+        Completes the multipart upload session.
+
+        When there exists no uploads in the session or the session completion
+        request fails unexpectedly due to client error, this method will
+        automatically abort the multipart upload session and raises
+        corresponding exceptions.
+
+        This method will raises :class:`EmptyMultipartUploadError` if no
+        parts are uploaded.
+
+        This method will raises :class:`UploadPartSizeOutOfBoundError` if an
+        uploaded part is too small. The minimum size of non-final part is
+        defined as ``AWS_S3_DEFAULT_MINIMUM_PART_SIZE`` (quantifier: bytes,
+        according to the AWS documentation) in the same module.
+
+        This method will raises :class:`UnexpectedMultipartUploadAbort` if an
+        uploaded part is too small. The minimum size of non-final part is
+        defined as ``AWS_S3_DEFAULT_MINIMUM_PART_SIZE`` (quantifier: bytes,
+        according to the AWS documentation) in the same module.
+        """
+        if not self.is_active:
+            return
+
+        if not self.parts:
+            self.abort()
+            raise EmptyMultipartUploadError(f'{self.bucket_name}/{self.object_key}')
+
+        for _ in as_completed(self.futures):
```

Comment:
You need to check on the status of each future. What if any of them raised an exception?